lib/renderer/api/web-frame.js

```diff
@@ -1,13 +1,67 @@
 'use strict'
 
 const { EventEmitter } = require('events')
-const { webFrame, WebFrame } = process.atomBinding('web_frame')
+const binding = process.atomBinding('web_frame')
 
-// WebFrame is an EventEmitter.
-Object.setPrototypeOf(WebFrame.prototype, EventEmitter.prototype)
-EventEmitter.call(webFrame)
+class WebFrame extends EventEmitter {
+  constructor (context) {
+    super()
 
-// Lots of webview would subscribe to webFrame's events.
-webFrame.setMaxListeners(0)
+    this.context = context
+    // Lots of webview would subscribe to webFrame's events.
+    this.setMaxListeners(0)
+  }
 
-module.exports = webFrame
+  findFrameByRoutingId (...args) {
+    return getWebFrame(binding._findFrameByRoutingId(this.context, ...args))
+  }
+
+  getFrameForSelector (...args) {
+    return getWebFrame(binding._getFrameForSelector(this.context, ...args))
+  }
+
+  findFrameByName (...args) {
+    return getWebFrame(binding._findFrameByName(this.context, ...args))
+  }
+
+  get opener () {
+    return getWebFrame(binding._getOpener(this.context))
+  }
+
+  get parent () {
+    return getWebFrame(binding._getParent(this.context))
+  }
+
+  get top () {
+    return getWebFrame(binding._getTop(this.context))
+  }
+
+  get firstChild () {
+    return getWebFrame(binding._getFirstChild(this.context))
+  }
+
+  get nextSibling () {
+    return getWebFrame(binding._getNextSibling(this.context))
+  }
+
+  get routingId () {
+    return binding._getRoutingId()
+  }
+}
+
+// Populate the methods.
+for (const name in binding) {
+  if (!name.startsWith('_')) { // some methods are manully populated above
+    WebFrame.prototype[name] = function (...args) {
+      return binding[name](this.context, ...args)
+    }
+  }
+}
+
+// Helper to return WebFrame or null depending on context.
+// TODO(zcbenz): Consider returning same WebFrame for the same context.
```

Comment:
Maybe using `gin::PerContextData` would help here, which is similar to `WebContentsUserData` in behavior.

Reply:
After thinking more about it, I realize the `blink::WebFrame` can change its v8 context when reloading/navigating, so we can not really map a v8 context to `WebFrame`.

To return same `WebFrame` for same frame, we probably need to use web elements as keys.

But no one has ever complained about it for so long, I should probably just remove this TODO.